# Portfolio Part 3 - Analysis of Loan Approval Data (2024 S2)

## Task background
In this Portfolio task, you will work on a new dataset named 'Loan Approval' which is a modified version from a synthetic Dataset for Risk Assessment and Loan Approval Modeling (many thanks to LORENZO ZOPPELLETTO for the sharing of this dataset). This dataset comprises 20,000 records of personal and financial data, designed to facilitate the development of predictive models for risk assessment and loan approval. In this portfolio part, you are mainly required to train classification models to determine the outcome of loan approval, indicating whether an applicant is likely to be approved or denied for a loan.

The dataset includes diverse features such as demographic information, credit history, employment status, income levels, existing debt, and other relevant financial metrics, providing a comprehensive foundation for sophisticated data-driven analysis and decision-making.

The dataset includes the following columns:

|Column|Meaning|
|:-----|:-----|
|ApplicationDate| Loan application date|
|Age| Applicant's age|
|AnnualIncome| Yearly income|
|CreditScore| Creditworthiness score|
|EmploymentStatus| Job situation|
|EducationLevel| Highest education attained|
|Experience| Work experience|
|LoanAmount| Requested loan size|
|LoanDuration| Loan repayment period|
|MaritalStatus| Applicant's marital state|
|NumberOfDependents| Number of dependents|
|HomeOwnershipStatus| Homeownership type|
|MonthlyDebtPayments| Monthly debt obligations|
|CreditCardUtilizationRate| Credit card usage percentage|
|NumberOfOpenCreditLines| Active credit lines|
|NumberOfCreditInquiries| Credit checks count|
|DebtToIncomeRatio| Debt to income proportion|
|BankruptcyHistory| Bankruptcy records|
|LoanPurpose| Reason for loan|
|PreviousLoanDefaults| Prior loan defaults|
|PaymentHistory| Past payment behavior|
|LengthOfCreditHistory| Credit history duration|
|SavingsAccountBalance| Savings account amount|
|CheckingAccountBalance| Checking account funds|
|TotalAssets| Total owned assets|
|TotalLiabilities| Total owed debts|
|MonthlyIncome| Income per month|
|UtilityBillsPaymentHistory| Utility payment record|
|JobTenure| Job duration|
|NetWorth| Total financial worth|
|BaseInterestRate| Starting interest rate|
|InterestRate| Applied interest rate|
|MonthlyLoanPayment| Monthly loan payment|
|TotalDebtToIncomeRatio| Total debt against income|
|LoanApproved| Loan approval status|
|RiskScore| Risk assessment score|

## Task Description
Your high level goal in this notebook is to build and evaluate predictive models for 'loan approval' from other available features. More specifically, you need to complete the following major tasks:

1. Clean and preprocess the dataset for the downstream data analysis tasks.

2. Build and evaluate logistic regression models with this datasets.

3. Build and evaluate KNN models with this datasets.

Note 1: While the main steps of each task have been given with the requirements, you should learn how to properly organise and comment your notebook by yourself to ensure that your notebook file is professional and readable.

Note 2: You will be evaluated on the accuracy of the model, the process that you produce the results,  and your clear description and justification of your implementation. So, try your best to comment your source code to showing your understanding and critical thinking.


## Load the dataset and show the basic information

In [9]:
import pandas as pd

data = pd.read_csv('loan_approval.csv')
data.head()

ApplicationDate   Age  AnnualIncome  CreditScore EmploymentStatus  \
0      2018-01-01  45.0         39948          617         Employed   
1      2018-01-02  38.0         39709          628         Employed   
2      2018-01-03  47.0         40724          570         Employed   
3      2018-01-04  58.0         69084          545         Employed   
4      2018-01-05  37.0        103264          594         Employed   

  EducationLevel  Experience  LoanAmount  LoanDuration MaritalStatus  ...  \
0         Master          22       13152            48       Married  ...   
1      Associate          15       26045            48        Single  ...   
2       Bachelor          26       17627            36       Married  ...   
3    High School          34       37898            96        Single  ...   
4      Associate          17        9184            36       Married  ...   

   MonthlyIncome UtilityBillsPaymentHistory  JobTenure  NetWorth  \
0    3329.000000                   0.724972         11    126928   
1    3309.083333                   0.935132          3     43609   
2    3393.666667                   0.872241          6      5205   
3    5757.000000                   0.896155          5     99452   
4    8605.333333                   0.941369          5    227019   

   BaseInterestRate  InterestRate  MonthlyLoanPayment  TotalDebtToIncomeRatio  \
0          0.199652      0.227590          419.805992                0.181077   
1          0.207045      0.201077          794.054238                0.389852   
2          0.217627      0.212548          666.406688                0.462157   
3          0.300398      0.300911         1047.506980                0.313098   
4          0.197184      0.175990          330.179140                0.070210   

  LoanApproved  RiskScore  
0            0        NaN  
1            0       52.0  
2            0        NaN  
3            0        NaN  
4            1        NaN  

[5 rows x 36 columns]

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 36 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ApplicationDate             20000 non-null  object 
 1   Age                         19900 non-null  float64
 2   AnnualIncome                20000 non-null  int64  
 3   CreditScore                 20000 non-null  int64  
 4   EmploymentStatus            20000 non-null  object 
 5   EducationLevel              20000 non-null  object 
 6   Experience                  20000 non-null  int64  
 7   LoanAmount                  20000 non-null  int64  
 8   LoanDuration                20000 non-null  int64  
 9   MaritalStatus               19900 non-null  object 
 10  NumberOfDependents          20000 non-null  int64  
 11  HomeOwnershipStatus         20000 non-null  object 
 12  MonthlyDebtPayments         20000 non-null  int64  
 13  CreditCardUtilizationRate   200

In [13]:
data.describe(include="all")

ApplicationDate           Age   AnnualIncome   CreditScore  \
count            20000  19900.000000   20000.000000  20000.000000   
unique           20000           NaN            NaN           NaN   
top         2018-01-01           NaN            NaN           NaN   
freq                 1           NaN            NaN           NaN   
mean               NaN     39.751759   59161.473550    571.612400   
std                NaN     11.630809   40350.845168     50.997358   
min                NaN     18.000000   15000.000000    343.000000   
25%                NaN     31.750000   31679.000000    540.000000   
50%                NaN     40.000000   48566.000000    578.000000   
75%                NaN     48.000000   74391.000000    609.000000   
max                NaN     80.000000  485341.000000    712.000000   

       EmploymentStatus EducationLevel    Experience     LoanAmount  \
count             20000          20000  20000.000000   20000.000000   
unique                3              5           NaN            NaN   
top            Employed       Bachelor           NaN            NaN   
freq              17036           6054           NaN            NaN   
mean                NaN            NaN     17.522750   24882.867800   
std                 NaN            NaN     11.316836   13427.421217   
min                 NaN            NaN      0.000000    3674.000000   
25%                 NaN            NaN      9.000000   15575.000000   
50%                 NaN            NaN     17.000000   21914.500000   
75%                 NaN            NaN     25.000000   30835.000000   
max                 NaN            NaN     61.000000  184732.000000   

        LoanDuration MaritalStatus  ...  MonthlyIncome  \
count   20000.000000         19900  ...   20000.000000   
unique           NaN             4  ...            NaN   
top              NaN       Married  ...            NaN   
freq             NaN          9999  ...            NaN   
mean       54.057000           NaN  ...    4891.715521   
std        24.664857           NaN  ...    3296.771598   
min        12.000000           NaN  ...    1250.000000   
25%        36.000000           NaN  ...    2629.583333   
50%        48.000000           NaN  ...    4034.750000   
75%        72.000000           NaN  ...    6163.000000   
max       120.000000           NaN  ...   25000.000000   

       UtilityBillsPaymentHistory     JobTenure      NetWorth  \
count                20000.000000  20000.000000  2.000000e+04   
unique                        NaN           NaN           NaN   
top                           NaN           NaN           NaN   
freq                          NaN           NaN           NaN   
mean                     0.799918      5.002650  7.229432e+04   
std                      0.120665      2.236804  1.179200e+05   
min                      0.259203      0.000000  1.000000e+03   
25%                      0.727379      3.000000  8.734750e+03   
50%                      0.820962      5.000000  3.285550e+04   
75%                      0.892333      6.000000  8.882550e+04   
max                      0.999433     16.000000  2.603208e+06   

        BaseInterestRate  InterestRate  MonthlyLoanPayment  \
count       20000.000000  20000.000000        20000.000000   
unique               NaN           NaN                 NaN   
top                  NaN           NaN                 NaN   
freq                 NaN           NaN                 NaN   
mean            0.239124      0.239110          911.607052   
std             0.035509      0.042205          674.583473   
min             0.130101      0.113310           97.030193   
25%             0.213889      0.209142          493.763700   
50%             0.236157      0.235390          728.511452   
75%             0.261533      0.265532         1112.770759   
max             0.405029      0.446787        10892.629520   

        TotalDebtToIncomeRatio  LoanApproved    RiskScore  
count             20000.000000  20000.0000

## Task 1: Clean the datasets (10 marks)

### Step 1.1 Handle the missing values with follwoing rules (5 marks)
1. If over 50% of the values of a column are missing, the column should be removed from the data frame; 
2. For a categorical column, if a row contains a missing value, you need to delete the whole row; 
3. For a numerical column, if a row contains a missing value, you need to perform a missing value imputation with the average value of the column.

### Step 1.2 Handle categorical attributes (5 marks)
1. If all the categorical values of a column are unique, this column does not provide any statistical informaiton and should be deleted.
2. Use one hot encoding to convert the categorical values into numerical ones.

## Task 2: Build a logistic regression classification model (25 marks)

### Step 2.1 Specify the features and the label, and split the dataset into training data and testing data (5 marks)
1. The attirbute "LoanApproved" is the label, which is the prediction target. The remaining attributes are the features.
2. The ratio for splitting the dataset is 80% for training and 20% for testing. Note that you need to set the "random_state" parameter as your student ID to produce your personlised splitting. Failing to do so will lose marks.

### Step 2.2 Build a logistic regression model (10 marks)
1. Train a logistic regression model
2. Report two classification performance metrics (accuracy and f1-score) on the testing data
3. Also report the two metrics on the training data, and compare the results with that of the testing data. Make a justification on whether the model is overfitting based on the comparison.

### Step 2.3 Perform the recursive feature elimination (RFE) technique to identify the effective features for building the model (10 marks)
1. Visulise the change of the two performance metrics with respect to the number of eliminated features using a line chart.
2. In terms of the visualisation result, select a good value for the number of eliminated features with considering both performance maximisation and feature minimisation (two competing goals). Run the RFE again with the chosen number of eliminated features to obtain the corresponding set of retained features.

## Task 3: Build a KNN classification model (25 marks)

### Step 3.1 Build 1-NN classifier (5 marks)
1. Slect the features identifed in Step 2.3 for this task
2. Buid 1-NN classifier and report two classification performance metrics (accuracy and f1-score) on the testing data
3. Also report the two metrics on the training data, and compare the results with that of the testing data. Make a justification on whether the model is overfitting based on the comparison.

### Step 3.2 Use the grid search and cross validation techniques to study the performance change with respect to the hyperparameter K (10 marks)
1. User grid search to search K in the range (1, 30) both inclusive with 5-fold cross validation. The performance metric used for search is accuracy.
2. Visualise the performance change with respect to K using a line chart. Report the two performance metrics for the best case.

### Step 3.3 Study how the distance metrics affect the model performance (10 marks)
1. Change the distance metric parameter to 3 distance types: 'euclidean'(also l2), 'l1', and 'cosine', respectively, and visualise the model performance with these 3 distances, using a bar chart for both accuracy and f1 scores.
2. Compare the performance metrics, which is the best? Which is the worest?